lib/line-numbers.js

```diff
@@ -1,93 +1,117 @@
 'use strict';
 
 const callsites = require('callsites');
+const micromatch = require('micromatch');
 const sourceMapSupport = require('source-map-support');
 const flatten = require('lodash/flatten');
 const parseTestSourceInFile = require('./test-ast');
 
-module.exports = {
-	parseLineNumbers,
-	parseLineNumbersInPath,
-	parseLineNumbersInPaths,
-	hasLineNumbersSuffix,
-	stripLineNumbersSuffix,
-	resolveEndLineNumberForTestInFile,
-	getLineNumberRangeForTestInFile,
-	isTestSelectedByLineNumbers
-};
-
 const NUMBER_REGEX = /^-?\d+$/;
 const RANGE_REGEX = /^(?<startGroup>-?\d+)-(?<endGroup>-?\d+)$/;
 const LINE_NUMBERS_REGEX = /^(?:-?\d+(?:-\d+)?,?)+$/;
-const PATH_DELIMITER = ':';
+const DELIMITER = ':';
+
+const pluralizeIf = (string, condition) => condition ? `${string}s` : string;
+
+class FormatError extends Error {
+	constructor(offender) {
+		const numberPluralized = pluralizeIf('number', offender && !NUMBER_REGEX.test(offender) && offender.length !== 1);
+		super(`Invalid line ${numberPluralized}: \`${offender}\`. Expected comma-separated list of \`[X|Y-Z]\`.`);
+		this.name = 'FormatError';
+	}
+}
 
 const parseNumber = string => parseInt(string, 10);
-const range = (start, end) => new Array(end - start + 1).fill(start).map((e, i) => e + i);
-const distinctArray = array => [...new Set(array)];
 const removeAllWhitespace = string => string.replace(/\s/g, '');
-const pluralizeIf = (string, condition) => condition ? `${string}s` : string;
-const getLineNumbersSuffix = path => path.split(PATH_DELIMITER).pop();
-const sortNumbersAscending = array => {
-	const sorted = [...array];
-	sorted.sort((a, b) => a - b);
-	return sorted;
-};
 
-function parseLineNumbers(string) {
-	if (!string) {
+const parseLineNumbers = suffix => {
+	const retval = [];
+	for (const string of suffix.split(',')) {
+		const trimmed = removeAllWhitespace(string);
+
+		if (NUMBER_REGEX.test(trimmed)) {
+			const number = parseNumber(trimmed);
+			if (number <= 0) {
+				throw new Error(`Invalid line number: \`${trimmed}\`. Line numbers must be positive.`);
+			}
+
+			retval.push(number);
+			continue;
+		}
+
+		const match = RANGE_REGEX.exec(trimmed);
+		if (match) {
+			const {startGroup, endGroup} = match.groups;
+			const start = parseNumber(startGroup);
+			const end = parseNumber(endGroup);
+			if (start < 0 || end < 0) {
+				throw new Error(`Invalid line number range: \`${trimmed}\`. Line numbers must be positive.`);
+			}
+
+			if (start > end) {
+				throw new Error(`Invalid line number range: \`${trimmed}\`. \`start\` must be lesser than \`end\`.`);
+			}
+
+			for (let i = start; i <= end; i++) {
+				retval.push(i);
+			}
+
+			continue;
+		}
+
 		throw new FormatError(string);
 	}
 
-	const parts = string.split(',');
-	const lineNumbers = flatten(parts.map(parsePart));
-	return sortNumbersAscending(distinctArray(lineNumbers));
-}
+	return retval;
```

Comment:
I genuinely think this should be replaced by/reverted to: `flatten(suffix.split(',').map(parsePart))`. Breaking out `parsePart()` is not strictly necessary. Inlining inside the `map()` works too. I just find it easier to read (and test!).

Reply:
Apologies… it's been three months now so I don't quite recall why I rewrote it like this. Possibly because I struggled to read the previous version of the code 😊

Reply:
OK, good. Apology accepted. I guess I owe you an apology back then for writing such un-readable code...